File: hamiltonianPath.cpp
#include<bits/stdc++.h>
using namespace std;

class Edge {
    public:
    int src;
    int nbr;
    int wt;

    Edge(int src , int nbr , int wt){
        this->src = src;
        this->wt = wt;
        this->nbr = nbr;
    }
};

void hamiltonian(vector<Edge> graph[] ,int vtces ,  int src , set<int>&visited , string psf , int osrc){

    if(visited.size() == vtces - 1){
        cout << psf ;
        bool closingEdgeFound = false;
        for(Edge edge : graph[src]){
            if(edge.nbr == osrc){
                closingEdgeFound = true;
            }
        }
        if(closingEdgeFound){
            cout << "*" << endl;
        }else{
            cout << "." << endl;
        }
        return;
    }

    visited.insert(src);
    for(Edge edge : graph[src]){
        if(visited.find(edge.nbr) == visited.end()){ // key not found
            hamiltonian(graph, vtces , edge.nbr , visited , psf + to_string(edge.nbr) , osrc);
        }
    }
    visited.erase(src);
}

int main() {
    int vtces;
    cin >> vtces;
    vector<Edge> graph[vtces];
    int e;
    cin >> e;
    for(int i = 0 ; i < e ; i++){
        int v1 , v2 , wt;
        cin >> v1 >> v2 >> wt;

        graph[v1].push_back(Edge(v1 , v2 , wt));
        graph[v2].push_back(Edge(v2 , v1 , wt));
    }
    int src;
    cin >> src;

    set<int> visited;
    hamiltonian(graph , vtces ,  src , visited , to_string(src) + "" , src);
    return 0;
}

File: hamiltonianPath.java
import java.io.*;
import java.util.*;

public class hamiltonianPath {
   static class Edge {
      int src;
      int nbr;
      int wt;

      Edge(int src, int nbr, int wt) {
         this.src = src;
         this.nbr = nbr;
         this.wt = wt;
      }
   }

   public static void hamiltonian(ArrayList<Edge>[] graph , int src , HashSet<Integer> visited , String psf , int osrc){
       if(visited.size() == graph.length - 1){
           System.out.print(psf);
            boolean closingEdgeFound = false;
            for(Edge e : graph[src]){
                if(e.nbr == osrc){
                    closingEdgeFound = true;
                }
            }
            if(closingEdgeFound == true){
                System.out.println("*");
            }
            else{
                System.out.println(".");
            }
            return;
       }
       visited.add(src);
       for(Edge edge : graph[src]){
           if(visited.contains(edge.nbr) == false){
               hamiltonian(graph , edge.nbr , visited , psf + edge.nbr , osrc);
           }
       }
       visited.remove(src);
   }


   public static void main(String[] args) throws Exception {
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

      int vtces = Integer.parseInt(br.readLine());
      ArrayList<Edge>[] graph = new ArrayList[vtces];
      for (int i = 0; i < vtces; i++) {
         graph[i] = new ArrayList<>();
      }

      int edges = Integer.parseInt(br.readLine());
      for (int i = 0; i < edges; i++) {
         String[] parts = br.readLine().split(" ");
         int v1 = Integer.parseInt(parts[0]);
         int v2 = Integer.parseInt(parts[1]);
         int wt = Integer.parseInt(parts[2]);
         graph[v1].add(new Edge(v1, v2, wt));
         graph[v2].add(new Edge(v2, v1, wt));
      }

      int src = Integer.parseInt(br.readLine());
    
      HashSet<Integer>visited = new HashSet<>();
      hamiltonian(graph , src , visited , src + "" , src);
       
   }


}